pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,280 @@
+		"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+	)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("initializing APIBindings for workspace")
+
+	// Start with the ClusterWorkspaceType specified by the ClusterWorkspace
+	leafCWT, err := b.getClusterWorkspaceType(logicalcluster.New(clusterWorkspace.Spec.Type.Path), string(clusterWorkspace.Spec.Type.Name))
+	if err != nil {
+		logger.Error(err, "error getting ClusterWorkspaceType")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error getting ClusterWorkspaceType %s|%s: %v",
+			clusterWorkspace.Spec.Type.Path, clusterWorkspace.Spec.Type.Name,
+			err,
+		)
+
+		return nil
+	}
+
+	// Get all the transitive ClusterWorkspaceTypes
+	cwts, err := b.transitiveTypeResolver.Resolve(leafCWT)
+	if err != nil {
+		logger.Error(err, "error resolving transitive types")
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			WorkspaceAPIBindingsInitialized,
+			WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of which APIBindings have been created for which APIExports
+	exportToBinding := map[apisv1alpha1.WorkspaceExportReference]*apisv1alpha1.APIBinding{}
+
+	for i := range bindings {
+		binding := bindings[i]
+
+		if binding.Spec.Reference.Workspace == nil {
+			continue
+		}
+
+		// Track what we have ("actual")
+		exportToBinding[*binding.Spec.Reference.Workspace] = binding
+	}
+
+	requiredExportRefs := map[tenancyv1alpha1.APIExportReference]struct{}{}
+
+	for _, cwt := range cwts {
+		logger := logging.WithObject(logger, cwt)
+		logger.V(2).Info("attempting to initialize APIBindings")
+
+		for i := range cwt.Spec.DefaultAPIBindings {
+			exportRef := cwt.Spec.DefaultAPIBindings[i]
+
+			// Keep track of unique set of expected exports across all CWTs
+			requiredExportRefs[exportRef] = struct{}{}
+
+			logger := logger.WithValues("apiExport.path", exportRef.Path, "apiExport.name", exportRef.Name)
+			ctx := klog.NewContext(ctx, logger)
+
+			// Make sure the APIExport exists
+			apiExport, err := b.getAPIExport(logicalcluster.New(exportRef.Path), exportRef.Name)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+
+			apiBindingName := generateAPIBindingName(clusterName, exportRef.Path, exportRef.Name)
+			logger = logger.WithValues("apiBindingName", apiBindingName)
+
+			if _, err = b.getAPIBinding(clusterName, apiBindingName); err == nil {
+				logger.V(4).Info("APIBinding already exists - skipping creation")
```

Comment:
If you're checking for AlreadyExists anyway, is this just to save one network hop?

Reply:
This is a cached read, so no network hop.

I've had users complain in the past about audit logs show a bunch of already-exists errors when a controller repeatedly tried to create rather than checking for existence first.

Reply:
Didn't know users would care about that - good to keep in mind for the future.